src/object/srv_ec_aggregate.c

```diff
@@ -560,31 +806,779 @@ agg_update_vos(struct ec_agg_entry *entry, bool write_parity)
+	struct ec_agg_param	*agg_param;
+	crt_opcode_t		 opcode;
+	crt_rpc_t		*rpc;
+	int			 rc = 0;
+
+	tgt_ep.ep_rank = entry->ae_peer_rank;
+	tgt_ep.ep_tag = entry->ae_peer_idx + 1;
+	opcode = DAOS_RPC_OPCODE(DAOS_OBJ_RPC_EC_AGGREGATE, DAOS_OBJ_MODULE,
+				 DAOS_OBJ_VERSION);
+	rc = crt_req_create(dss_get_module_info()->dmi_ctx, &tgt_ep, opcode,
+			    &rpc);
+	if (rc) {
+		D_ERROR("crt_req_create failed: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+	ec_agg_in = crt_req_get(rpc);
+	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
+	uuid_copy(ec_agg_in->ea_pool_uuid,
+		  agg_param->ap_pool_info.api_pool_uuid);
+	uuid_copy(ec_agg_in->ea_coh_uuid,
+		  agg_param->ap_pool_info.api_poh_uuid);
+	uuid_copy(ec_agg_in->ea_cont_uuid,
+		  agg_param->ap_pool_info.api_cont_uuid);
+	uuid_copy(ec_agg_in->ea_coh_uuid,
+		  agg_param->ap_pool_info.api_coh_uuid);
+	ec_agg_in->ea_oid = entry->ae_oid;
+	ec_agg_in->ea_oid.id_shard--;
+	ec_agg_in->ea_dkey = entry->ae_dkey;
+	ec_agg_in->ea_akey = entry->ae_akey;
+	ec_agg_in->ea_epoch_range.epr_lo = agg_param->ap_epr.epr_lo;
+	ec_agg_in->ea_epoch_range.epr_hi = entry->ae_cur_stripe.as_hi_epoch;
+	ec_agg_in->ea_stripenum = entry->ae_cur_stripe.as_stripenum;
+	ec_agg_in->ea_map_ver =
+		agg_param->ap_pool_info.api_pool->sp_map_version;
+
+	if (stripe_ud->asu_write_par) {
+		ec_agg_in->ea_rsize = entry->ae_rsize;
+		buf = (unsigned char *)
+			entry->ae_sgl.sg_iovs[AGG_IOV_PARITY].iov_buf;
+		d_iov_set(&iov,
+			  &buf[entry->ae_oca->u.ec.e_len * entry->ae_rsize],
+			  entry->ae_oca->u.ec.e_len * entry->ae_rsize);
+		sgl.sg_iovs = &iov;
+		sgl.sg_nr = sgl.sg_nr_out = 1;
+		rc = crt_bulk_create(dss_get_module_info()->dmi_ctx, &sgl,
+				     CRT_BULK_RW, &ec_agg_in->ea_bulk);
+		if (rc) {
+			D_ERROR("crt_bulk_create failed: "DF_RC"\n", DP_RC(rc));
+			goto out;
+		}
+	}
+
+	if (entry->ae_cur_stripe.as_ho_ext_cnt || !agg_contained(entry)) {
+		struct ec_agg_extent	*ext;
+		int			 i = 0;
+
+		ec_agg_in->ea_remove_nr = entry->ae_cur_stripe.as_ho_ext_cnt +
+			entry->ae_cur_stripe.as_extent_cnt;
+		D_ALLOC_ARRAY(ec_agg_in->ea_remove_recxs.ca_arrays,
+			      ec_agg_in->ea_remove_nr);
+		if (ec_agg_in->ea_remove_recxs.ca_arrays == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+		D_ALLOC_ARRAY(ec_agg_in->ea_remove_eps.ca_arrays,
+			      ec_agg_in->ea_remove_nr);
+		if (ec_agg_in->ea_remove_eps.ca_arrays == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+
+		d_list_for_each_entry(ext, &entry->ae_cur_stripe.as_dextents,
+				      ae_link) {
+			ec_agg_in->ea_remove_recxs.ca_arrays[i] =
+				ext->ae_orig_recx;
+			ec_agg_in->ea_remove_eps.ca_arrays[i++] =
+				ext->ae_epoch;
+		}
+		if (entry->ae_cur_stripe.as_ho_ext_cnt) {
+			d_list_for_each_entry(ext,
+					      &entry->
+					      ae_cur_stripe.as_hoextents,
+					      ae_link) {
+				ec_agg_in->ea_remove_recxs.ca_arrays[i] =
+					ext->ae_orig_recx;
+				ec_agg_in->ea_remove_eps.ca_arrays[i++] =
+					ext->ae_epoch;
+			}
+		}
+	}
```

Comment:
Not sure I follow here.  I thought the peer parity node can calculate these remove extents/epoch by itself? since the rpc includes parity extent and epoch?